[jit] Support indexing with type besides string and int

Eg.
```python
import thunder

class Foo:
    def __hash__(self):
        return id(self)

f = Foo()
d = {f: 1, 1: 0}

def foo():
    return d[f]

jfoo = thunder.jit(foo)
jfoo()
```

Error
```python
File "/home/kkalambarkar/lightning-thunder/thunder/__init__.py", line 508, in get_computation_and_inputs
    jit_results: TraceResults = thunder_general_jit(
  File "/home/kkalambarkar/lightning-thunder/thunder/core/jit_ext.py", line 1755, in thunder_general_jit
    pro_to_comp_proxies, pro_to_epi_proxies = unpack_inputs(ctx, prologue_trace, pro_to_comp, pro_to_epi, args, kwargs)
  File "/home/kkalambarkar/lightning-thunder/thunder/core/jit_ext.py", line 1549, in unpack_inputs
    unpack(a)
  File "/home/kkalambarkar/lightning-thunder/thunder/core/jit_ext.py", line 1499, in unpack
    raise NotImplementedError(f"Exception occured unpacking object from {p.history}") from e
NotImplementedError: Exception occured unpacking object from ProvenanceRecord(
  i1 = INPUT_FN()
  i2 = LOAD_ATTR(i1, '__globals__')
  i3 = BINARY_SUBSCR(i2, 'd')
  i4 = BINARY_SUBSCR(i2, 'f')
  i5 = BINARY_SUBSCR(i3, i4)
)
```

This shows up when trying to support tracing through `torch.autocast` which calls `find_spec` which ends up calling [StashKey](https://docs.pytest.org/en/stable/reference/reference.html#pytest.StashKey) from pytest (when running with pytest).

Repro with Pytest 
NOTE - Run below script with Pytest
```python
import thunder
from importlib.util import find_spec

def test():
    def foo():
        return find_spec("some_mod")

    jfoo = thunder.jit(foo)
    jfoo()
```

Fails with
```python
E               NotImplementedError: Exception occured unpacking object from ProvenanceRecord(
E                 i1 = INPUT_FN()
E                 i2 = LOAD_ATTR(i1, '__globals__')
E                 i3 = BINARY_SUBSCR(i2, 'find_spec')
E                 i4 = LOAD_ATTR(i3, '__globals__')
E                 i5 = BINARY_SUBSCR(i4, 'sys')
E                 i6 = LOAD_ATTR(i5, 'meta_path')
E                 i7 = BINARY_SUBSCR(i6, 0)
E                 i8 = LOAD_ATTR(i7, 'config')
E                 i9 = LOAD_ATTR(i8, 'stash')
E                 i10 = LOAD_ATTR(i9, '_storage')
E                 i11 = LOAD_ATTR(i7, 'find_spec')
E                 i12 = LOAD_ATTR(i11, '__func__')
E                 i13 = LOAD_ATTR(i12, '__globals__')
E                 i14 = BINARY_SUBSCR(i13, 'assertstate_key')
E                 i15 = BINARY_SUBSCR(i10, i14)
E                 i16 = LOAD_ATTR(i15, 'trace')
E                 i17 = LOAD_ATTR(i16, 'tags')
E                 i18 = BINARY_SUBSCR(i17, 0)
E               )
```

Ref in PyTorch -
https://github.com/pytorch/pytorch/blob/8c3ab21490f336ee8fcc19b8047bd89e8ce21ea5/torch/amp/autocast_mode.py#L261-L265
https://github.com/pytorch/pytorch/blob/8c3ab21490f336ee8fcc19b8047bd89e8ce21ea5/torch/cuda/amp/common.py#L11-L12

cc: @t-vi

--- a/thunder/core/jit_ext.py
+++ b/thunder/core/jit_ext.py
@@ -1384,7 +1384,7 @@ def unpack_inputs(ctx, prologue_trace, pro_to_comp_inps, pro_to_epi_inps, args, 
                 output = Proxy("subscr")  # name? collectify?
             else:
                 output = p
-            if isinstance(idx, (int, str)):
+            if isinstance(idx, (int, str, Proxy)):
                 if isinstance(idx, int):
                     idx = int(idx)
                 elif isinstance(idx, str):

--- a/thunder/tests/test_core.py
+++ b/thunder/tests/test_core.py
@@ -2937,3 +2937,32 @@ def test_return_bsym_has_none_output(requires_grad):
             return_bsym = trace.bound_symbols[-1]
             assert return_bsym.sym.id == thunder.prims.PrimIDs.RETURN
             assert return_bsym.output is None
+
+
+def test_indexing_with_hashable_object():
+    class HashableClass:
+        def __hash__(self):
+            return id(self)
+
+    h = HashableClass()
+    d = {h: 1, 1: 0}
+
+    def fn():
+        return d[h]
+
+    jfn = thunder.jit(fn)
+    assert jfn() == 1
+    assert thunder.cache_misses(jfn) == 1  # Due to first compilation.
+
+    # Call jfn with no changes
+    # this should be cache hit.
+    assert jfn() == 1
+    assert thunder.cache_hits(jfn) == 1
+    assert thunder.cache_misses(jfn) == 1
+
+    # Change the value of the captured dict.
+    # This should be a cache miss, verify that.
+    d[h] = 2
+    assert jfn() == 2  # Verify that jfn now returns 2
+    assert thunder.cache_hits(jfn) == 1
+    assert thunder.cache_misses(jfn) == 2

